File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] ret=new int[nums.length];
        int prod=1;
        int zer=0;
        for (int i:nums) {
            if(i==0){
                zer++;
                continue;
            }
            prod*=i;
        }
        if (zer==0) {
            for (int j=0;j<ret.length;j++) {
                ret[j]=prod/nums[j];
            }            
        }else if(zer==1){
            for (int j=0;j<ret.length;j++) {
                ret[j]=(nums[j]!=0)?0:prod;
            }
        }else{
            for (int j=0;j<ret.length;j++) {
                ret[j]=0;
            }
        }
        return ret;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums):
        n=len(nums)
        ret=[1]*n
        lft_prod=1
        rit_prod=1
        for i in range(n):
            ret[i]*=lft_prod
            lft_prod*=nums[i]            
            ret[n-1-i]*=rit_prod
            rit_prod*=nums[n-1-i]
        return ret